ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryQueue.java

```diff
@@ -0,0 +1,197 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.query;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.physical.pull.PullQueryRow;
+import io.confluent.ksql.util.KeyValue;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class PullQueryQueue implements BlockingRowQueue {
+
+  public static final int BLOCKING_QUEUE_CAPACITY = 50;
+  private static final Function<PullQueryRow, KeyValue<List<?>, GenericRow>>
+      PQ_ROW_TO_KEY_VALUE = row -> KeyValue.keyValue(null, row.getGenericRow());
+
+  private final BlockingQueue<PullQueryRow> rowQueue;
+  private final int offerTimeoutMs;
+  private volatile boolean closed = false;
+
+  private LimitHandler limitHandler;
+  private LimitQueueCallback callback;
+
+  public PullQueryQueue() {
+    this(BLOCKING_QUEUE_CAPACITY, 100);
+  }
+
+  public PullQueryQueue(
+      final int queueSizeLimit,
+      final int offerTimeoutMs) {
+    this.callback = new QueueToCompletionCallback(this::isClosed);
+    this.rowQueue = new LinkedBlockingQueue<>(queueSizeLimit);
+    this.offerTimeoutMs = offerTimeoutMs;
+    this.limitHandler = () -> { };
+  }
+
+  @Override
+  public void setLimitHandler(final LimitHandler limitHandler) {
+    this.limitHandler = limitHandler;
+    callback.setLimitHandler(limitHandler);
+  }
+
+  @Override
+  public void setQueuedCallback(final Runnable queuedCallback) {
+    final LimitQueueCallback parent = callback;
+
+    callback = new LimitQueueCallback() {
+      @Override
+      public boolean shouldQueue() {
+        return parent.shouldQueue();
+      }
+
+      @Override
+      public void onQueued() {
+        parent.onQueued();
+        queuedCallback.run();
+      }
+
+      @Override
+      public void setLimitHandler(final LimitHandler limitHandler) {
+        parent.setLimitHandler(limitHandler);
+      }
+    };
+  }
+
+  @Override
+  public KeyValue<List<?>, GenericRow> poll(final long timeout, final TimeUnit unit)
+      throws InterruptedException {
+    return Optional.ofNullable(rowQueue.poll(timeout, unit))
+        .map(PQ_ROW_TO_KEY_VALUE)
+        .orElse(null);
+  }
+
+  @Override
+  public KeyValue<List<?>, GenericRow> poll() {
+    return Optional.ofNullable(rowQueue.poll())
+        .map(PQ_ROW_TO_KEY_VALUE)
+        .orElse(null);
+  }
+
+  @Override
+  public void drainTo(final Collection<? super KeyValue<List<?>, GenericRow>> collection) {
+    final List<PullQueryRow> list = new ArrayList<>();
```

Comment:
can we avoid iterating all the elements in the queue twice? seems unnecessary

Reply:
I could repeatedly call poll to drain it, though this is meant to be more efficient with all of the synchronization.  In the end, the array is only 50 elements long, max.

I think it's probably better to take the small allocation rather than the synchronization hit, but maybe you disagree. I can change it, if you prefer.

Reply:
ack - now that I understand the motivation we can leave it as is. in the future we can profile and if this ends up being a hotspot we can look into it then